tests/src/python/providertestbase.py

```diff
@@ -415,6 +415,21 @@ def testFeatureCount(self):
             self.assertEqual(count, 0)
             self.assertEqual(self.source.featureCount(), 5)
 
+    def testEmpty(self):
+        self.assertFalse(self.source.empty())
+        self.assertFalse(self.source.emptyUnknown())
+
+        if self.source.supportsSubsetString():
+            # Add a subset string and test feature count
+            subset = self.getSubsetString()
+            self.source.setSubsetString(subset)
+            self.assertFalse(self.source.empty())
+            subsetNoMatching = getSubsetStringNoMatching(self)
+            self.source.setSubsetString(subsetNoMatching)
+            self.assertTrue(self.source.empty())
+            self.source.setSubsetString(None)
+            self.assertFalse(self.source.empty())
+
```

Comment:
You could also add a test like testTruncate from providertestbase.py - i.e. get an editable layer (if supported by the provider's test), delete all features (by calling truncate), and then test if the layer returns true for isEmpty(). This would directly test the provider's implementation without the subset strings coming into play at all.